Bug suspected in Microsoft API when enabling storage threat protection - error parsing URI

<!---
Please note the following potential times when an issue might be in Terraform core:

* [Configuration Language](https://www.terraform.io/docs/configuration/index.html) or resource ordering issues
* [State](https://www.terraform.io/docs/state/index.html) and [State Backend](https://www.terraform.io/docs/backends/index.html) issues
* [Provisioner](https://www.terraform.io/docs/provisioners/index.html) issues
* [Registry](https://registry.terraform.io/) issues
* Spans resources across multiple providers

If you are running into one of these scenarios, we recommend opening an issue in the [Terraform core repository](https://github.com/hashicorp/terraform/) instead.
--->

<!--- Please keep this note for the community --->

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

<!--- Thank you for keeping this note for the community --->

### Terraform (and AzureRM Provider) Version

<!--- Please run `terraform -v` to show the Terraform core version and provider version(s). If you are not running the latest version of Terraform or the provider, please upgrade because your issue may have already been fixed. [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions). --->

### Affected Resource(s)

<!--- Please list the affected resources and data sources. --->

Terraform v0.12.30
provider.azurerm v2.42.0
### Terraform Configuration Files

<!--- Information about code formatting: https://help.github.com/articles/basic-writing-and-formatting-syntax/#quoting-code --->

```hcl
resource "azurerm_advanced_threat_protection" "business" {
  target_resource_id = azurerm_storage_account.business.id
  enabled            = true
}
```

### Debug Output

<!---
Please provide a link to a GitHub Gist containing the complete debug output. Please do NOT paste the debug output in the issue; just paste a link to the Gist.

To obtain the debug output, see the [Terraform documentation on debugging](https://www.terraform.io/docs/internals/debugging.html).
--->

### Panic Output

<!--- If Terraform produced a panic, please provide a link to a GitHub Gist containing the output of the `crash.log`. --->

### Expected Behaviour

Advanced Threat Protection should successfully be enabled for the given storage account

### Actual Behaviour
Terraform errored with:

Error: [ERROR] Unable to parse Advanced Threat Protection Set ID "subscriptions/ece6...c45/resourceGroups/test_xyz_shared/providers/Microsoft.Storage/storageAccounts/xyzbusiness/providers/Microsoft.Security/advancedThreatProtectionSettings/current": Cannot parse Azure ID: parse "subscriptions/ece6...c45/resourceGroups/test_xyz_shared/providers/Microsoft.Storage/storageAccounts/xyzbusiness/providers/Microsoft.Security/advancedThreatProtectionSettings/current": invalid URI for request


<!--- What actually happened? --->

### Steps to Reproduce

<!--- Please list the steps required to reproduce the issue. --->

1. `terraform apply`

### Important Factoids

State file showed the resource as tainted and looking closely we can see the leading '/' is missing on the id (ignore the ...):

      "mode": "managed",
      "type": "azurerm_advanced_threat_protection",
      "name": "business",
      "provider": "provider.azurerm",
      "instances": [
        {
          "status": "tainted",
        ...
            "id": "subscriptions/ece6...c45/resourceGroups/..../providers/Microsoft.Storage/storageAccounts/..../providers/Microsoft.Security/advancedThreatProtectionSettings/current",
            "target_resource_id": "/subscriptions/ece6...45/resourceGroups/..../providers/Microsoft.Storage/storageAccounts/...",
...
    },

Updating the state file by adding the leading '/', re-running the apply and it works.

Noticed that the unit test in resourceid_test.go checks for missing leading '/' so can only assume that the Microsoft API is broken.
Whilst this may not be a terraform provider bug, the community here should be aware and also might be in a position to help.

### References

<!---
Information about referencing Github Issues: https://help.github.com/articles/basic-writing-and-formatting-syntax/#referencing-issues-and-pull-requests

Are there any other GitHub issues (open or closed) or pull requests that should be linked here? Such as vendor documentation?
--->

--- a/azurerm/internal/services/securitycenter/advanced_threat_protection_resource.go
+++ b/azurerm/internal/services/securitycenter/advanced_threat_protection_resource.go
@@ -11,6 +11,7 @@ import (
 	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/tf"
 	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/clients"
 	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/securitycenter/parse"
+	azSchema "github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/tf/schema"
 	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/timeouts"
 	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/utils"
 )
@@ -22,9 +23,10 @@ func resourceAdvancedThreatProtection() *schema.Resource {
 		Update: resourceAdvancedThreatProtectionCreateUpdate,
 		Delete: resourceAdvancedThreatProtectionDelete,
 
-		Importer: &schema.ResourceImporter{
-			State: schema.ImportStatePassthrough,
-		},
+		Importer: azSchema.ValidateResourceIDPriorToImport(func(id string) error {
+			_, err := parse.AdvancedThreatProtectionID(id)
+			return err
+		}),
 
 		Timeouts: &schema.ResourceTimeout{
 			Create: schema.DefaultTimeout(30 * time.Minute),

--- a/azurerm/internal/services/securitycenter/advanced_threat_protection_resource.go
+++ b/azurerm/internal/services/securitycenter/advanced_threat_protection_resource.go
@@ -10,6 +10,7 @@ import (
 	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/azure"
 	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/tf"
 	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/clients"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/securitycenter/migration"
 	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/securitycenter/parse"
 	azSchema "github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/tf/schema"
 	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/timeouts"
@@ -27,6 +28,11 @@ func resourceAdvancedThreatProtection() *schema.Resource {
 			_, err := parse.AdvancedThreatProtectionID(id)
 			return err
 		}),
+
+		SchemaVersion: 1,
+		StateUpgraders: []schema.StateUpgrader{
+			migration.AdvancedThreatProtectionV0ToV1(),
+		},
 
 		Timeouts: &schema.ResourceTimeout{
 			Create: schema.DefaultTimeout(30 * time.Minute),

--- a/azurerm/internal/services/securitycenter/migration/advanced_threat_protection_v0_to_v1.go
+++ b/azurerm/internal/services/securitycenter/migration/advanced_threat_protection_v0_to_v1.go
@@ -1,0 +1,54 @@
+package migration
+
+import (
+	"fmt"
+	"log"
+	"strings"
+
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/securitycenter/parse"
+)
+
+func AdvancedThreatProtectionV0ToV1() schema.StateUpgrader {
+	return schema.StateUpgrader{
+		Version: 0,
+		Type:    advancedThreatProtectionV0Schema().CoreConfigSchema().ImpliedType(),
+		Upgrade: advancedThreadProtectionV0toV1Upgrade,
+	}
+}
+
+func advancedThreatProtectionV0Schema() *schema.Resource {
+	return &schema.Resource{
+		Schema: map[string]*schema.Schema{
+			"target_resource_id": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"enabled": {
+				Type:     schema.TypeBool,
+				Required: true,
+			},
+		},
+	}
+}
+
+func advancedThreadProtectionV0toV1Upgrade(rawState map[string]interface{}, meta interface{}) (map[string]interface{}, error) {
+	oldId := rawState["id"].(string)
+
+	// remove the existing `/` if it's present (2.42+) which'll do nothing if it wasn't (2.38)
+	newId := strings.TrimPrefix(oldId, "/")
+	newId = fmt.Sprintf("/%s", oldId)
+
+	parsedId, err := parse.AdvancedThreatProtectionID(newId)
+	if err != nil {
+		return nil, err
+	}
+
+	newId = parsedId.ID()
+
+	log.Printf("[DEBUG] Updating ID from %q to %q", oldId, newId)
+	rawState["id"] = newId
+	return rawState, nil
+}

--- a/azurerm/internal/services/securitycenter/migration/advanced_threat_protection_v0_to_v1.go
+++ b/azurerm/internal/services/securitycenter/migration/advanced_threat_protection_v0_to_v1.go
@@ -38,8 +38,7 @@ func advancedThreadProtectionV0toV1Upgrade(rawState map[string]interface{}, meta
 	oldId := rawState["id"].(string)
 
 	// remove the existing `/` if it's present (2.42+) which'll do nothing if it wasn't (2.38)
-	newId := strings.TrimPrefix(oldId, "/")
-	newId = fmt.Sprintf("/%s", oldId)
+	newId := fmt.Sprintf("/%s", strings.TrimPrefix(oldId, "/"))
 
 	parsedId, err := parse.AdvancedThreatProtectionID(newId)
 	if err != nil {

